### PR TITLE
Add logical_not as an optional callback in backend

### PR DESCRIPTION
**Pull request by howard0su**

### CODE REVIEW

File: nx/lib/nx.ex [hunk 1]
```diff
@@ -4246,10 +4246,20 @@ defmodule Nx do
...
-    out = %T{shape: {}, type: type, names: []}
-    zero = Nx.BinaryBackend.from_binary(out, number_to_binary(0, type), [])
-    element_wise_pred_op(tensor, zero, :equal)
+    output = Nx.template(tensor.shape, {:u, 8}, names: tensor.names)
+
+    Nx.Shared.optional(:logical_not, [tensor], output, fn tensor ->
+      type = tensor.type
+
+      zero =
+        Nx.BinaryBackend.from_binary(
```
Comment (polvalente) [inline nx/lib/nx.ex:4255]: I don't think this should be using the binary backend. It should be a backend-agnostic implementation.

Perhaps some form of Nx.broadcast can help with this

File: nx/lib/nx.ex [hunk 2]
```diff
@@ -4246,10 +4246,20 @@ defmodule Nx do
...
+          %T{shape: {}, type: type, names: []},
+          number_to_binary(0, type),
+          []
+        )
+
+      element_wise_pred_op(tensor, zero, :equal)
```
Comment (polvalente) [inline nx/lib/nx.ex:4261]: By the way, can't you just pass 0 instead of a scalar tensor?

File: nx/lib/nx.ex (hunk 1 above)
Comment (josevalim) [reply nx/lib/nx.ex:4255]: It is fine to use the BinaryBackend here, we are basically creating a constant. I understand it is confusing though!

File: nx/lib/nx.ex (hunk 2 above)
Comment (josevalim) [reply nx/lib/nx.ex:4261]: 0 will lead to a different type coercion logic, that's why we do the dance above!

**Review by josevalim — APPROVED** [review]

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**